protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,509 @@
+                                relay_peer_id,
+                                dst_addr,
+                                dst_peer_id,
+                                send_back: tx,
+                            })
+                            .await?;
+                        let stream = rx.await?.map_err(|()| RelayError::Connect)?;
+                        Ok(stream)
+                    }
+                    .boxed(),
+                ))
+            }
+        }
+    }
+
+    fn address_translation(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.inner_transport.address_translation(server, observed)
+    }
+}
+
+#[derive(Default)]
+struct RelayedMultiaddr {
+    relay_peer_id: Option<PeerId>,
+    relay_addr: Option<Multiaddr>,
+    dst_peer_id: Option<PeerId>,
+    dst_addr: Option<Multiaddr>,
+}
+
+/// Parse a [`Multiaddr`] containing a [`Protocol::P2pCircuit`].
+///
+/// Returns `Ok(Err(provided_addr))` when passed address contains no [`Protocol::P2pCircuit`].
+///
+/// Returns `Err(_)` when address is malformed.
+fn parse_relayed_multiaddr(
+    addr: Multiaddr,
+) -> Result<Result<RelayedMultiaddr, Multiaddr>, RelayError> {
+    if !addr.iter().any(|p| matches!(p, Protocol::P2pCircuit)) {
+        return Ok(Err(addr));
+    }
+
+    let mut relayed_multiaddr = RelayedMultiaddr::default();
+
+    let mut before_circuit = true;
+    for protocol in addr.into_iter() {
+        match protocol {
+            Protocol::P2pCircuit => {
+                if before_circuit {
+                    before_circuit = false;
+                } else {
+                    return Err(RelayError::MultipleCircuitRelayProtocolsUnsupported);
+                }
+            }
+            Protocol::P2p(hash) => {
+                let peer_id = PeerId::from_multihash(hash).map_err(|_| RelayError::InvalidHash)?;
+
+                if before_circuit {
+                    if relayed_multiaddr.relay_peer_id.is_some() {
+                        return Err(RelayError::MalformedMultiaddr);
+                    }
+                    relayed_multiaddr.relay_peer_id = Some(peer_id)
+                } else {
+                    if relayed_multiaddr.dst_peer_id.is_some() {
+                        return Err(RelayError::MalformedMultiaddr);
+                    }
+                    relayed_multiaddr.dst_peer_id = Some(peer_id)
+                }
+            }
+            p => {
+                if before_circuit {
+                    relayed_multiaddr
+                        .relay_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                } else {
+                    relayed_multiaddr
+                        .dst_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                }
+            }
+        }
+    }
+
+    Ok(Ok(relayed_multiaddr))
+}
+
+#[pin_project(project = RelayListenerProj)]
```

Comment:
I haven't looked into it in detail, but why not use `EitherListenStream` here?

Reply:
Again, good catch. Fixed via a8f21ae7.